common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java

```diff
@@ -65,104 +70,182 @@ public static void submitSilent(@NonNull final BaseCommand command){
                 }
 
-                try {
-                    //Try executing request
-                    result = command.execute();
-                } catch (final Exception e) {
-                    //Capture any resulting exception and map to MsalException type
-                    Logger.errorPII(
+                //Check cache to see if the same command completed in the last 30 seconds
+                commandResult = sCommandResultCache.get(command);
+
+                //If nothing in cache, execute the command and cache the result
+                if(commandResult == null) {
+                    commandResult = executeCommand(command);
+                    cacheCommandResult(command, commandResult);
+                }else{
+                    Logger.info(
                             TAG + methodName,
-                            "Silent request failed with Exception",
-                            e
+                            "Silent command result returned from cache."
                     );
-                    if (e instanceof BaseException) {
-                        baseException = (BaseException) e;
-                    } else {
-                        baseException = ExceptionAdapter.baseExceptionFromException(e);
-                    }
                 }
 
-                if (baseException != null) {
-                    //Post On Error
-                    final BaseException finalException = baseException;
-                    handler.post(new Runnable() {
-                        @Override
-                        public void run() {
-                            command.getCallback().onError(finalException);
-                        }
-                    });
-                } else {
-                    if(result != null && result instanceof AcquireTokenResult){
-                        //Handler handler, final BaseCommand command, BaseException baseException, AcquireTokenResult result
-                        processTokenResult(handler, command, baseException, (AcquireTokenResult)result );
-                    }else{
-                        //For commands that don't return an AcquireTokenResult
-                        final Object returnResult = result;
+                //Return the result via the callback
+                returnCommandResult(command, commandResult, handler);
 
-                        handler.post(new Runnable() {
-                            @Override
-                            public void run() {
-                                command.getCallback().onTaskCompleted(returnResult);
-                            }
-                        });
-                    }
-                }
+                sExecutingCommands.remove(command);
 
                 Telemetry.getInstance().flush(correlationId);
             }
         });
     }
 
+    public static void clearCommandCache(){
```

Comment:
Why is this method public? Anyone can call this method from anywhere, and if they are able do execute this as part of their loop, then I believe we won't throttle and we won't provide any protection against excessive requests.

I think this method can package-private or even private as it is not being used anywhere at the moment, other than the [command caching tests in MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/801/files#diff-721e4828ff2ace5339c66f849f90da31R289-R344).

Reply:
Let's discuss.  I want to be able to re-use your investments in the existing tests.  If package private or private this call will be inaccessible to the tests.  What's your suggestion on how to achieve the same thing?  I don't thing our goal in this case is prevent malicious apps from misusing the library... if anyone wants to attack the service they can by simply opening an http connection and making many requests.  Our  goal here is to protect the system against coding bugs in clients.  Unintentional misuse.  Note that many other things are in "internal" packages that we don't intend for clients to call directly.

Reply:
Ok.  I took this feedback and worked around accessibility by creating a helper class in the MSAL unit tests.